### ISSUE TITLE
Android pip upstreaming

### ISSUE THREAD

**Issue by mhsmith** [chaquo/chaquopy#981]
The main patches are:
* Recognizing the `android` platform tag, so that e.g. building for an `android_25` platform will still accept wheels tagged as `android_21`.
* Preferring older native wheels over newer sdists. I left a comment in the pip patch explains why we don't simply use `--prefer-binary`, but I'm not sure whether that still applies.
* Some custom error messages, which could probably be moved to the Gradle plugin or the pip_install script with only a small loss of detail.

Related:
* #980

**Comment by IsakTheHacker** [comment]
Will the pip upstreaming render the current [package repository](https://chaquo.com/pypi-13.1/) obsolete?

**Comment by mhsmith** [comment]
No, because it will be some time before Android wheels are accepted by PyPI, although that's something we're working on long-term. But I'd like to be able to use an unpatched version of pip to install from our existing repository, because the patches make it more difficult to upgrade pip to deal with issues like https://github.com/beeware/briefcase/issues/1480.

**Comment by mhsmith** [comment]
Android wheels are [now accepted by PyPI](https://beeware.org/mobile-wheels/).

**Comment by mhsmith** [comment]
> `sys_platform` and the others might be possible to handle with a cross venv

This turned out to be infeasible, because monkey patching `os.name`, `sys.platform` etc. affects too many things. Simulating iOS on macOS works well enough in Briefcase, and simulating Android on Linux would probably be fine as well. But simulating Android on macOS had some undesirable effects, like changing the location of the user cache directory, and simulating Android on Windows broke many more things, not just in pip but even in the standard library, as it tried to import Unix-specific modules that don't exist.

So we'll need to retain some patches for environment markers and `--only-binary`, but we should be able to remove the others.

**Comment by mhsmith** [comment]
> we're already setting a few environment variables like `CC` to nonexistent paths – maybe we could point them at actual scripts which would give a better error message

I started looking into this, and it should be feasible on Unix, but Windows is more awkward. In setuptools it looks like everything depends on the vcvarsall.bat script. Some environment variables which may be able to affect this are `DISTUTILS_USE_SDK`, `ProgramFiles` and `MSSdk` (the latter maybe only for older versions of Visual Studio used by obsolete versions of Python). 

Of course, other build systems will be different. I didn't investigate any of them, but based on [this article](https://labs.quansight.org/blog/pep-517-build-system-popularity), the next most popular ones are poetry, hatchling and flit.

I also didn’t look into languages other than C or C++, such as Rust. However, this isn’t a regression, as the stable version of Chaquopy doesn’t do anything about those languages either. I’ve seen reports that Chaquopy users who have the Rust development tools installed may get some distance through an sdist build, and maybe even appear to succeed, though the resulting wheel will of course be for the wrong platform. This is something we really want to avoid.

So rather than pursue this any further, I'll see if it's now feasible to use `--only-binary`. This would replace the wide range of confusing error messages from a failed sdist build, with a single “Could not find a version that satisfies the requirement” error, and it would fail much more quickly and consistently.

It may also enable pip's new solver to handle situations like a pure-Python `pydantic` with a 1-1 dependency on a native `pydantic-core`. If there is only one `pydantic-core` version available as an Android wheel, I think it will backtrack through the `pydantic` versions until it finds the one that uses it. Without `--only-binary`, I think it would select the sdist of the newest version, and then give up after failing to build it.

**Comment by mhsmith** [comment]
The main question is how many wheels we'd need to build to make this an acceptably small regression. I used the procedure in pypi/README-internal.md to try installing the top 1000 packages on PyPI with `--only-binary`, and compared this to Chaquopy 16.1.0, using Python 3.10 in both cases:

<table>
<tr><td>Successes with 16.1.0</td><td>780</td></tr>
<tr><td>Changed to failure with --only-binary</td><td>61</td></tr>
<tr><td>Changed to success with --only-binary</td><td>6</td></tr>
<tr><td>Successes with --only-binary</td><td>725</td></tr>
</table>

Those that changed to success are unlikely to be useful, because they're caused by reverting the package to a much older version that had a pure-Python wheel, sometimes even a prerelease:
```
2 snowflake-connector-python-1.9.1
1 fastavro-0.17.5
1 polars-runtime-32-0.0.0
1 pydantic-core-0.0.1
1 watchfiles-0.0.0a1
```

Of those that changed to failure, more than half are because of `markupsafe` and `pyyaml`, two packages with pure-Python fallbacks which we already listed as examples in chaquopy_monkey.py:
```
   7 MarkupSafe>=2.0
   6 markupsafe>=2.1.1
   5 pyyaml
   3 MarkupSafe>=2.1.1
   3 MarkupSafe>=0.9.2
   2 tornado>=6.2
   2 tornado>=6.1.0
   2 pyyaml>=5.1
   2 markupsafe
   1 watchdog>=5.0.0
   1 watchdog>=2.0
   1 watchdog
   1 tornado
   1 thrift==0.16.0
   1 thrift<0.21.0,>=0.16.0
   1 thrift
   1 setproctitle
   1 rapidfuzz<4.0.0,>=3.0.0
   1 rapidfuzz
   1 pyyaml>=5.4.1
   1 PyYAML>=5.3.1
   1 pyyaml>=5.3.1
   1 PyYAML>=5.1.0
   1 pyyaml>=3.12
   1 PyYAML<7.0,>=5.1
   1 PyYAML<6.1,>=3.10
   1 pymongo
   1 pure-sasl>=0.6.2
   1 oss2
   1 nvidia-cusparselt-cu12
   1 maxminddb<3.0.0,>=2.7.0
   1 maxminddb
   1 libclang
   1 html5validator
   1 hf-xet<2.0.0,>=1.2.0; platform_machine == "x86_64" or platform_machine == "amd64" or platform_machine == "AMD64" or platform_machine == "arm64" or platform_machine == "aarch64"
   1 docopt
   1 clickhouse-connect
   1 antlr4-python3-runtime==4.9.*
```

Like msgpack (#1299), these two projects  refuse to release a pure-Python wheel because that would cause people on platforms without wheels, or on prerelease versions of Python, who can currently build an accelerated version from the sdist, to switch to the slower pure-Python version:
* https://github.com/pallets/markupsafe/issues/483
* https://github.com/yaml/pyyaml/issues/95

**Comment by mhsmith** [comment]
I've now released Android wheels for `markupsafe`, `pyyaml` and `tornado`. Of the others:
* `thrift` is fairly popular, so we may build it if anyone requests it.
* `maxminddb` already has official Android wheels on PyPI for Python 3.13 and later. We may build wheels for 3.12 and older if anyone requests it.
* `watchdog` is actually pure-Python on every platform except macOS, so if we do build this, it should be a pure-Python wheel ([watchdog#1137 (comment)](https://github.com/gorakhargosh/watchdog/issues/1137#issuecomment-3541009044)).

**Comment by mhsmith** [comment]
After this PR is complete, I'll retest the top 1000 packages and record the results in #1427.
